File: examples/05-callbacks/main.ml
external twice : ('a -> unit) -> 'a -> unit = "caml_twice"
external twice_catch : ('a -> unit) -> 'a -> unit = "caml_twice_catch"

exception MyExn

let () =
  Callback.register_exception "myexn" MyExn

let () =
  twice print_endline "hello";
  (* with caml_callback, exceptions skip over the C code *)
  (try twice (fun _ -> raise Exit) 0
   with Exit -> print_endline "raised Exit");
  (* caml_callback_exn allows catching exceptions in the C code *)
  twice_catch (
    let r = ref false in
    fun _ ->
      if !r then raise Exit
      else (r := true; raise MyExn)
  ) 0

File: examples/05-callbacks/main.c
#include <stdio.h>
#define CAML_NAME_SPACE
#include <caml/mlvalues.h>
#include <caml/alloc.h>
#include <caml/memory.h>
#include <caml/callback.h>

value caml_twice(value f, value x) {
  CAMLparam2(f, x);
  caml_callback(f, x);
  caml_callback(f, x);
  CAMLreturn(Val_unit);
}

void handle_exn(value exn) {
  CAMLparam1(exn);
  if (exn == *caml_named_value("myexn"))
    printf("C: raised MyExn\n");
  else
    printf("C: raised an exception\n");
  CAMLreturn0;
}

value caml_twice_catch(value f, value x) {
  CAMLparam2(f, x);
  CAMLlocal1(ret);
  ret = caml_callback_exn(f, x);
  if(Is_exception_result(ret))
    handle_exn(Extract_exception(ret));
  ret = caml_callback_exn(f, x);
  if(Is_exception_result(ret))
    handle_exn(Extract_exception(ret));
  CAMLreturn(Val_unit);
}
